### PR TITLE
Jq/pushappend

### PR DESCRIPTION
**Pull request by quinnj**

### CODE REVIEW

File: src/array.jl [hunk 1]
```diff
@@ -193,3 +193,13 @@ function getindex(A::CatOrdArray, i::Int)
 end
 
 levels!(A::CatOrdArray, newlevels::Vector) = _levels!(A, newlevels)
+
+Base.push!{T, R}(A::CategoricalArray{T, 1, R}, item::T) = (push!(A.refs, get!(A.pool, item)); return A)
```
Comment (nalimilan) [inline src/array.jl]: Better use the `CatOrdArray` alias as elsewhere, rather than duplicating definitions.

**Comment by nalimilan** [comment]
I'm afraid these definitions won't work in two cases:
- when the new values are not included in existing levels
- when a null value is added (for nullable arrays only)

For `push!`, the solution is to resize the array, and use `setindex!`. For `append!`, you need to call `levels!(res, intersect(levels(a), levels(b))` first.

**Comment by quinnj** [comment]
Would we really want `intersect` instead of `union`?

**Comment by quinnj** [comment]
Another method I realized I need is `empty!(A)`. Do you think for CategoricalArrays, it would be proper to just `empty!(A.refs)`? Or should we empty the pool too?

**Comment by nalimilan** [comment]
Sorry, I meant `union` of course. As regards `empty!`, I'm really not sure what's the best solution. Let's preserve the levels for now, at least until we find a reason to act differently.

**Comment by quinnj** [comment]
Alrighty, cleaned up and tests passing. Might not be the most performant solutions in the world, but they seem to be functional.

**Comment by quinnj** [comment]
I do find it interesting that there are two coverage systems on CI, but no travis or appveyor? :)

File: test/11_array.jl [hunk 2]
```diff
@@ -132,6 +132,38 @@ for (A, V, M) in ((NominalArray, NominalVector, NominalMatrix),
         @test x[3] === x.pool.valindex[1]
         @test levels(x) == ["e", "a", "b", "c"]
 
+        x = V{String, R}(a)
+        push!(x, "a")
+        @test length(x) == 4
+        @test String(x[end]) == "a"
```
Comment (nalimilan) [inline test/11_array.jl]: No need for `String` here, `==` is implemented to allow direct comparisons with `NominalValue`.

File: test/11_array.jl [hunk 3]
```diff
@@ -132,6 +132,38 @@ for (A, V, M) in ((NominalArray, NominalVector, NominalMatrix),
...
+
+        push!(x, "zz")
+        @test length(x) == 5
+        @test String(x[end]) == "zz"
```
Comment (nalimilan) [inline test/11_array.jl]: Same here (and in 12_nullablearray.jl).

File: test/11_array.jl [hunk 4]
```diff
@@ -132,6 +132,38 @@ for (A, V, M) in ((NominalArray, NominalVector, NominalMatrix),
         @test x[3] === x.pool.valindex[1]
         @test levels(x) == ["e", "a", "b", "c"]
 
+        x = V{String, R}(a)
```
Comment (nalimilan) [inline test/11_array.jl]: Why not use the same `x` as above? That's the approach I've taken, which is more likely to discover subtle corruption bugs than simple operations on newly-created arrays.

File: test/11_array.jl [hunk 5]
```diff
@@ -132,6 +132,38 @@ for (A, V, M) in ((NominalArray, NominalVector, NominalMatrix),
...
+        append!(x, y)
+        @test length(x) == 9
+
+        a = ["a","b","c"]
+        x = V{String, R}(a)
+        a = ["z","y","x"]
+        y = V{String, R}(a)
+        append!(x, y)
+        @test length(x) == 6
+        @test String(x[1]) == "a"
```
Comment (nalimilan) [inline test/11_array.jl]: Same here.

File: test/11_array.jl [hunk 6]
```diff
@@ -132,6 +132,38 @@ for (A, V, M) in ((NominalArray, NominalVector, NominalMatrix),
...
+        @test String(x[2]) == "b"
+        @test String(x[3]) == "c"
+        @test String(x[4]) == "z"
+        @test String(x[5]) == "y"
+        @test String(x[6]) == "x"
+
+        empty!(x)
+        @test length(x) == 0
```
Comment (nalimilan) [inline test/11_array.jl]: Could you also test `levels` here and for other functions? A particularly important case is when `x` and `y` share some levels, but not in the same order. Should be easy to add to your existing tests.

**Comment by nalimilan** [comment]
> Alrighty, cleaned up and tests passing. Might not be the most performant solutions in the world, but they seem to be functional.

Thanks. These look reasonably fast. We could be slightly faster by 1) copying the integer codes for `x` as-is, and 2) computing a map of level codes in `y` to level codes in the new vector, storing it in a vector, and accessing it to recode instead of relying on `setindex!`. But let's leave these for later.

> I do find it interesting that there are two coverage systems on CI, but no travis or appveyor? :)

AFAIK Travis is enabled. We could add AppVeyor, but nothing should be Windows-specific in this package, so not a high priority.

**Comment by quinnj** [comment]
Ok updated.

I mention travis because it doesn't seem to be showing up here to show its status.

File: test/11_array.jl [hunk 7]
```diff
@@ -132,6 +132,37 @@ for (A, V, M) in ((NominalArray, NominalVector, NominalMatrix),
...
+        append!(x, y)
+        @test length(x) == 12
+        @test x[end-2] == "z"
+        @test x[end-1] == "y"
+        @test x[end] == "x"
+        @test levels(x) == ["e", "a", "b", "c", "zz", "z", "y", "x"]
+
+
+        empty!(x)
+        @test length(x) == 0
```
Comment (nalimilan) [inline test/11_array.jl]: Please test that levels are preserved.

File: test/11_array.jl [hunk 8]
```diff
@@ -132,6 +132,37 @@ for (A, V, M) in ((NominalArray, NominalVector, NominalMatrix),
...
+        b = ["z","y","x"]
+        y = V{String, R}(b)
+        append!(x, y)
+        @test length(x) == 12
+        @test x[end-2] == "z"
+        @test x[end-1] == "y"
+        @test x[end] == "x"
+        @test levels(x) == ["e", "a", "b", "c", "zz", "z", "y", "x"]
+
+
```
Comment (nalimilan) [inline test/11_array.jl]: One line break is enough.

**Comment by nalimilan** [comment]
There's a test failure on Travis. Apart from this and the two details I noted, should be OK.

**Comment by quinnj** [comment]
Weird, I can't reproduce the error when I run the test file by itself, only when I run the entire test suite. Can you think of any reason I'd see that? Unfortunately, the backtraces seemed to be messed up when the test suite is run in full and it doesn't tell me the individual test that is failing. I'll keep trying ot figure out what's going on here.

File: test/12_nullablearray.jl [hunk 9]
```diff
@@ -145,6 +145,44 @@ for (A, V, M) in ((NullableNominalArray, NullableNominalVector, NullableNominalM
...
+            @test levels(x) == ["e", "c", "zz"]
+
+            push!(x, eltype(x)())
+            @test length(x) == 7
+            @test isnull(x[end])
+            @test levels(x) == ["e", "c", "zz"]
+
+            y = V{String, R}(a)
+            append!(x, y)
+            @test length(x) == 10
```
Comment (nalimilan) [inline test/12_nullablearray.jl]: Can you also test the contents of `x` after appending? Same below.

**Comment by nalimilan** [comment]
Got it. The error was only happening with `--bounds-check=yes`, since it comes from indexing inside an `@inbounds` loop in `_levels!` at `src/array.jl:167`. So the new tests actually uncovered a real bug in `levels!` triggered by `append!`: the `x > 0` condition should apply to both lines in the loop. Please make that change, even removing the intermediate variable `j` which isn't really useful.

Could you also repeat the new test code for each series of tests in each file? The test are currently repeated (with some variations were needed) for vectors (without then with nulls) and matrices (without then with nulls). I could probably simplify this, but for now I'd prefer that you follow that structure.

**Comment by quinnj** [comment]
Well, we won't need the tests for matrices since we're only dealing with vector operations here (push!, append!)

**Comment by nalimilan** [comment]
Right, so please just repeat the tests twice (one for without nullables, then one with nullables).

**Comment by quinnj** [comment]
Ok, think we're good to go here.

File: test/11_array.jl [hunk 10]
```diff
@@ -132,6 +132,49 @@ for (A, V, M) in ((NominalArray, NominalVector, NominalMatrix),
...
+        push!(x, "zz")
+        @test length(x) == 5
+        @test x[end] == "zz"
+        @test levels(x) == ["e", "a", "b", "c", "zz"]
+
+        push!(x, x[1])
+        @test length(x) == 6
+        @test x[1] == x[end]
+        @test levels(x) == ["e", "a", "b", "c", "zz"]
+        # ["c", "a", "a", "a", "zz", "c"]
```
Comment (nalimilan) [inline test/11_array.jl]: Leftover?

File: test/11_array.jl [hunk 11]
```diff
@@ -132,6 +132,49 @@ for (A, V, M) in ((NominalArray, NominalVector, NominalMatrix),
...
+
+        append!(x, x)
+        @test length(x) == 12
+        @test x[1] == "c"
```
Comment (nalimilan) [inline test/11_array.jl]: Can't this be replaced with `x == ["c", "a", ...]`?

File: test/11_array.jl [hunk 12]
```diff
@@ -132,6 +132,49 @@ for (A, V, M) in ((NominalArray, NominalVector, NominalMatrix),
...
+        @test x[9] == "a"
+        @test x[10] == "a"
+        @test x[11] == "zz"
+        @test x[12] == "c"
+
+        b = ["z","y","x"]
+        y = V{String, R}(b)
+        append!(x, y)
+        @test length(x) == 15
+        @test x[end-2] == "z"
```
Comment (nalimilan) [inline test/11_array.jl]: Doing `x == "c", "a", ...]` would be good to check the full contents of `x`.

File: test/11_array.jl [hunk 13]
```diff
@@ -240,6 +283,43 @@ for (A, V, M) in ((NominalArray, NominalVector, NominalMatrix),
         @test x[4] === x.pool.valindex[4]
         @test levels(x) == vcat(unique(a), -1)
 
+        push!(x, 2.0)
```
Comment (nalimilan) [inline test/11_array.jl]: Same remarks for this series of tests.

File: test/12_nullablearray.jl [hunk 14]
```diff
@@ -145,6 +145,45 @@ for (A, V, M) in ((NullableNominalArray, NullableNominalVector, NullableNominalM
...
+            @test levels(x) == ["e", "c", "zz"]
+
+            push!(x, eltype(x)())
+            @test length(x) == 7
+            @test isnull(x[end])
+            @test levels(x) == ["e", "c", "zz"]
+
+            y = V{String, R}(levels(x))
+            append!(x, y)
+            @test length(x) == 10
```
Comment (nalimilan) [inline test/12_nullablearray.jl]: Here too, you can check the full contents of `x` using `isequal(x, NullableArray(["e", "c", ...]))`.

**Comment by quinnj** [comment]
Alright, let's see what CI says about this.

**Comment by nalimilan** [comment]
Thanks! Sorry it took so many rounds of review!

Tell me if you need me to tag a new release.

**Comment by quinnj** [comment]
No worries. Yeah, let's tag a new release. I can do that if you'd like.

**Comment by nalimilan** [comment]
I'd just like to include commit https://github.com/nalimilan/CategoricalArrays.jl/commit/d4a4e70136320cbbb34d05942a4992f925ec2ce1, but tests unexplicably fail on Travis while they pass locally. Could you checkout `nl/test` and tell me whether tests pass on your machine?

**Comment by nalimilan** [comment]
Ah, nevermind, I just realized this depends on https://github.com/JuliaStats/NullableArrays.jl/pull/141, which isn't merged upstream. Let me find a temporary fix.

**Comment by nalimilan** [comment]
OK, should be good as soon as all tests are done. Feel free to tag the new commit as version 0.0.3 (I'm going to be offline for a few hours).

**Comment by quinnj** [comment]
Do you want to do a "release" on github and I can just use that?

**Comment by nalimilan** [comment]
Ah, right. See https://github.com/nalimilan/CategoricalArrays.jl/releases/tag/v0.0.3

**Pull request merged**